src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -0,0 +1,198 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+#nullable enable
+namespace Microsoft.CodeAnalysis
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0016:Add public types and members to the declared API", Justification = "In Progress")]
+    public abstract class GeneratorDriver
+    {
+        protected readonly GeneratorDriverState _state;
+
+        protected GeneratorDriver(GeneratorDriverState state)
+        {
+            _state = state;
+        }
+
+        protected GeneratorDriver(Compilation compilation, ParseOptions parseOptions)
+        {
+            _state = new GeneratorDriverState(compilation, parseOptions, ImmutableArray<GeneratorProvider>.Empty, ImmutableArray<AdditionalText>.Empty, ImmutableArray<PendingEdit>.Empty, ImmutableDictionary<GeneratorProvider, ImmutableArray<GeneratedSourceText>>.Empty, finalCompilation: null, editsFailed: true);
+        }
+
+        public GeneratorDriver RunFullGeneration(Compilation compilation, out Compilation outputCompilation, CancellationToken cancellationToken = default)
```

Comment:
>GeneratorDriver RunFullGeneration [](start = 15, length = 33)

I am probably missing something or confused by the names. For me the name "GeneratorDriver" means this is an object that drives generation, i.e. it does some work and perhaps returns a result. The fact that it also creates another driver feels strange to me. We already done the work, what is there left to drive?

Reply:
In the compiler case, there is nothing left to drive. In the IDE it returns a driver that represents the latest 'ongoing' state, which can be used to drive the next iteration.

---
In reply to: [382836760](https://github.com/dotnet/roslyn/pull/41789#discussion_r382836760) [](ancestors = 382836760)